glib/src/subclass/signal.rs

```diff
@@ -68,9 +70,254 @@ impl fmt::Debug for SignalInvocationHint {
+    ///
+    /// # Safety
+    ///
+    /// The caller has to ensure it's a valid signal identifier.
+    pub unsafe fn new(id: NonZeroU32) -> Self {
+        Self(id)
+    }
+
+    #[doc(alias = "g_signal_parse_name")]
+    pub fn parse_name(name: &str, type_: Type, force_detail: bool) -> Option<(Self, crate::Quark)> {
+        let mut signal_id = std::mem::MaybeUninit::uninit();
+        let mut detail_quark = std::mem::MaybeUninit::uninit();
+        unsafe {
+            let found: bool = from_glib(gobject_ffi::g_signal_parse_name(
+                name.to_glib_none().0,
+                type_.to_glib(),
+                signal_id.as_mut_ptr(),
+                detail_quark.as_mut_ptr(),
+                force_detail.to_glib(),
+            ));
+
+            if found {
+                Some((
+                    from_glib(signal_id.assume_init()),
+                    crate::Quark::from_glib(detail_quark.assume_init()),
+                ))
+            } else {
+                None
+            }
+        }
+    }
+
+    /// Find a SignalId by it's `name` and the `type` it connects to.
+    #[doc(alias = "g_signal_lookup")]
+    pub fn lookup(name: &str, type_: Type) -> Option<Self> {
+        unsafe {
+            let signal_id = gobject_ffi::g_signal_lookup(name.to_glib_none().0, type_.to_glib());
+            if signal_id == 0 {
+                None
+            } else {
+                Some(Self::new(NonZeroU32::new_unchecked(signal_id)))
+            }
+        }
+    }
+
+    /// Queries more in-depth information about the current signal.
+    #[doc(alias = "g_signal_query")]
+    pub fn query(&self) -> SignalQuery {
+        unsafe {
+            let mut query_ptr = std::mem::MaybeUninit::uninit();
+            gobject_ffi::g_signal_query(self.to_glib(), query_ptr.as_mut_ptr());
+            let query = query_ptr.assume_init();
+            assert_ne!(query.signal_id, 0);
+            SignalQuery(query)
+        }
+    }
+
+    /// Find the signal name.
+    #[doc(alias = "g_signal_name")]
+    pub fn name<'a>(&self) -> &'a str {
+        unsafe {
+            let ptr = gobject_ffi::g_signal_name(self.to_glib());
+            std::ffi::CStr::from_ptr(ptr).to_str().unwrap()
+        }
+    }
+}
+
+#[doc(hidden)]
+impl FromGlib<u32> for SignalId {
+    unsafe fn from_glib(signal_id: u32) -> Self {
+        assert_ne!(signal_id, 0);
+        Self::new(NonZeroU32::new_unchecked(signal_id))
+    }
+}
+
+#[doc(hidden)]
+impl ToGlib for SignalId {
+    type GlibType = u32;
+
+    fn to_glib(&self) -> u32 {
+        self.0.into()
+    }
+}
+
+impl PartialEq for SignalId {
+    fn eq(&self, other: &SignalId) -> bool {
+        self.0.eq(&other.0)
+    }
+}
+
+#[derive(Copy, Clone, Hash)]
+pub struct SignalType(ffi::GType);
+
+impl SignalType {
+    pub fn with_static_scope(type_: Type) -> Self {
+        Self(type_.to_glib() | gobject_ffi::G_TYPE_FLAG_RESERVED_ID_BIT)
+    }
+
+    pub fn static_scope(&self) -> bool {
```

Comment:
Maybe for completeness also add a `fn type(&self) -> Type`?

Also why doesn't clippy complain that `self` should be taken by value here? Should be the case here and also for `Type` everywhere as they're `Copy` and small.

Reply:
> Also why doesn't clippy complain that self should be taken by value here? Should be the case here and also for Type everywhere as they're Copy and small.

Maybe such linter doesn't exist because it doesn't complain for SignalId neither for example. Should I handle it on the same PR or it can be done in an another one?

Reply:
Well, it exists and I saw it triggered for such cases in the past :) That would be https://rust-lang.github.io/rust-clippy/master/index.html#trivially_copy_pass_by_ref

Can be done in a separate PR but then please create an issue so we don't forget.

Reply:
see https://github.com/gtk-rs/gtk-rs/issues/313